Wikipedia/Code/PageHistoryViewController.swift

```diff
@@ -0,0 +1,547 @@
+    }
+
+    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
+        super.traitCollectionDidChange(previousTraitCollection)
+        cellLayoutEstimate = nil
+    }
+
+    private func getPageHistory() {
+        isLoadingData = true
+
+        pageHistoryFetcher.fetchRevisionInfo(pageURL, requestParams: pageHistoryFetcherParams, failure: { error in
+            print(error)
+            self.isLoadingData = false
+        }) { results in
+            self.pageHistorySections.append(contentsOf: results.items())
+            self.pageHistoryFetcherParams = results.getPageHistoryRequestParameters(self.pageURL)
+            self.batchComplete = results.batchComplete()
+            self.isLoadingData = false
+            DispatchQueue.main.async {
+                self.collectionView.reloadData()
+            }
+        }
+    }
+
+    override func scrollViewDidScroll(_ scrollView: UIScrollView) {
+        super.scrollViewDidScroll(scrollView)
+        guard shouldLoadNewData else {
+            return
+        }
+        getPageHistory()
+    }
+
+    private enum State {
+        case idle
+        case editing
+    }
+
+    private var maxNumberOfRevisionsSelected: Bool {
+        assert((0...2).contains(selectedCellsCount))
+        return selectedCellsCount == 2
+    }
+    private var selectedCellsCount = 0
+
+    private var pageHistoryHintController: PageHistoryHintController? {
+        return hintController as? PageHistoryHintController
+    }
+
+    private var state: State = .idle {
+        didSet {
+            switch state {
+            case .idle:
+                selectedCellsCount = 0
+                pageHistoryHintController?.hide(true, presenter: self, theme: theme)
+                openSelectionIndex = 0
+                navigationItem.rightBarButtonItem = compareButton
+                collectionView.indexPathsForSelectedItems?.forEach { collectionView.deselectItem(at: $0, animated: true) }
+                forEachVisibleCell { (indexPath: IndexPath, cell: PageHistoryCollectionViewCell) in
+                    self.updateSelectionThemeModel(nil, for: cell, at: indexPath)
+                    self.collectionView.deselectItem(at: indexPath, animated: true)
+                    cell.enableEditing(true) // confusing, have a reset method
+                    cell.setEditing(false)
+                }
+                resetComparisonSelectionButtons()
+                navigationController?.setToolbarHidden(true, animated: true)
+            case .editing:
+                navigationItem.rightBarButtonItem = cancelComparisonButton
+                collectionView.allowsMultipleSelection = true
+                forEachVisibleCell { $1.setEditing(true) }
+                compareToolbarButton.isEnabled = false
+                NSLayoutConstraint.activate([
+                    firstComparisonSelectionButton.widthAnchor.constraint(equalToConstant: 90),
```

Comment:
I haven't confirmed but might want to make sure you remove these constraints when in .idle so you don't have multiple width constraints added.

Reply:
definitely 👍